### PR TITLE
Add new pagedQuery methods to GraphQLBehavior

### PR DESCRIPTION
**Pull request by richardmcclellan**
This PR replaces https://github.com/aws-amplify/amplify-android/pull/434

_Goal:_

- Expose pagination related properties so that datastore can query API and obtain all results.

_Description of changes:_

- Adds new `pagedQuery` method to GraphQLBehavior for obtaining the next set of results.
- Adds `Page` model, which encapsulates a GraphQLResponse, and knows if there is a nextPage (`hasNextPage()`), and how to get the next page (`queryNextPage(success, failure)`)

_To do before merging (but after team consensus on this approach):_

- Review this design and get consensus with iOS / JS
- Add integration test
- We don't actually need to add a new `pagedQuery` method to GraphQLBehavior, but it was easier for me to do it this way just to prove that this works.    Now that it does, we should just update the existing `query` methods which return a GraphQLResponse<Iterable<T>> to return a `Page<T>` instead.
 - Replace `Iterable` with `List`
 - Add Javadoc, address check-style issues.

_Customer Experience_

```
    private void listTodos() {
        try {
            int limit = 3;
            GraphQLRequest<Todo> request = AppSyncGraphQLRequestFactory.buildQuery(Todo.class, null, limit, null);
            Amplify.API.pagedQuery(request, this::handlePage, this::handleError);
        } catch(ApiException apiFailure) {
            handleError(apiFailure);
        }
    }

    private void handlePage(Page<Todo> page) {
        GraphQLResponse<Iterable<Todo>> response = page.getResponse();
        if (response.hasData()) {
            for (Todo todo : response.getData()) {
                Log.d(TAG, "todo: " + todo.getName());
            }
        }
        if (response.hasErrors()) {
            Log.d(TAG, "error: " + error);
        }

        if(page.hasNextPage()) {
            page.queryNextPage(this::handlePage, this::handleError);
        }
    }

    private void handleError(ApiException exception) {
        Log.e(TAG, exception.getMessage(), exception);
    }
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-api/src/main/java/com/amplifyframework/api/aws/AWSApiPlugin.java [hunk 1]
```diff
@@ -785,6 +821,32 @@ private String selectApiName(Set<String> apiClients) throws ApiException {
...
+                .client(clientDetails.okHttpClient())
+                .request(graphQLRequest)
+                .responseFactory(gqlResponseFactory)
+                .onResponse(onResponse)
+                .onFailure(onFailure)
+                .build();
+    }
+
+
+    private <T> PagedResultOperation<T> buildPageOperation(
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/AWSApiPlugin.java:843]: If it is only used once, does it make sense to inline it?

File: aws-api/src/main/java/com/amplifyframework/api/aws/PagedResultOperation.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+
+/**
+ * An operation to enqueue a GraphQL request to OkHttp client,
+ * with the goal of obtaining a list of responses. For example,
+ * this is used for a LIST query vs. a GET query or most mutations.
+ * @param <T> Casted type of GraphQL result data
+ */
+public final class PagedResultOperation<T> extends GraphQLOperation<T> {
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/PagedResultOperation.java:50]: As prior tech debt, we forked the single and multiple response operations. So, we had two things, that are mostly the same. Now, we'll have three. Seems like a good time to pause and try to figure out a way to have less code duplication between the three.

File: aws-api/src/main/java/com/amplifyframework/api/aws/PagedResultOperation.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    private static final Logger LOG = Amplify.Logging.forNamespace("amplify:aws-api");
+    private static final String CONTENT_TYPE = "application/json";
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/PagedResultOperation.java:52]: I think OkHttp has a constant for this, under `MediaType`, somewhere?

File: core/src/main/java/com/amplifyframework/api/GraphQlBehavior.java [hunk 4]
```diff
@@ -139,6 +140,29 @@
...
 
+    /**
+     * Perform a GraphQL query against a configured GraphQL endpoint.
+     * This operation is asynchronous and may be canceled by calling
+     * cancel on the returned operation. The response will be provided
+     * to the `onResponse` callback.  If there is data present
+     * in the response, it will be cast as the requested class type.
+     * Requires that only one API is configured in your
+     * `amplifyconfiguration.json`. Otherwise, emits an ApiException to
+     * the provided `onFailure` callback.
```
Comment (jamesonwilliams) [inline core/src/main/java/com/amplifyframework/api/GraphQlBehavior.java:151]: These new docs should start off with some strongly indicative language "PAges! How does a page work? You can use a page ... "

File: aws-api/src/main/java/com/amplifyframework/api/aws/AppSyncPage.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+import com.amplifyframework.api.ApiException;
+import com.amplifyframework.api.graphql.GraphQLOperation;
+import com.amplifyframework.api.graphql.GraphQLRequest;
+import com.amplifyframework.api.graphql.GraphQLResponse;
+import com.amplifyframework.api.graphql.Page;
+import com.amplifyframework.core.Amplify;
+import com.amplifyframework.core.Consumer;
+
+public class AppSyncPage<T> extends Page<T> {
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/AppSyncPage.java:13]: `final class`?

File: core/src/main/java/com/amplifyframework/api/graphql/Page.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    public abstract boolean hasNextPage();
+
+    public abstract GraphQLOperation<T> queryNextPage(@NonNull Consumer<Page<T>> onResponse,
+                                                      @NonNull Consumer<ApiException> onFailure);
+
+    protected Page(GraphQLResponse<Iterable<T>> response) {
+        this.response = response;
+    }
+
+    public GraphQLResponse<Iterable<T>> getResponse() {
```
Comment (jamesonwilliams) [inline core/src/main/java/com/amplifyframework/api/graphql/Page.java:20]: `final` to close the getter. Can this be just `protected`?

File: aws-api/src/main/java/com/amplifyframework/api/aws/PagedResultOperation.java (hunk 2 above)
Comment (richardmcclellan) [reply aws-api/src/main/java/com/amplifyframework/api/aws/PagedResultOperation.java:50]: Agreed.  Adding a 3rd thing was the easiest way for me to get something going, but with a few more changes, I can pretty easily just replace the multi response operation with the paged operation, so we are left with 2.  I'll try to do some investigation as well to see if we can reduce it even further.

File: aws-api/src/main/java/com/amplifyframework/api/aws/AppSyncPage.java (hunk 5 above)
Comment (richardmcclellan) [reply aws-api/src/main/java/com/amplifyframework/api/aws/AppSyncPage.java:13]: Addressed in https://github.com/aws-amplify/amplify-android/pull/488

File: core/src/main/java/com/amplifyframework/api/graphql/Page.java (hunk 6 above)
Comment (richardmcclellan) [reply core/src/main/java/com/amplifyframework/api/graphql/Page.java:20]: > `final` to close the getter. 

Addressed in https://github.com/aws-amplify/amplify-android/pull/488

> Can this be just `protected`?

No, this is designed to be exposed - this is the only way a customer would be able to access the `GraphQLResponse`, in order to access `data` and `errors`.

File: aws-api/src/main/java/com/amplifyframework/api/aws/PagedResultOperation.java (hunk 3 above)
Comment (richardmcclellan) [reply aws-api/src/main/java/com/amplifyframework/api/aws/PagedResultOperation.java:52]: Hmm, I couldn't find a constant for it.  The [example in the OkHttp docs](https://square.github.io/okhttp/) seems to do this the same way, with a hardcoded String.  Further down in this file, it looks like this String is converted to a MediaType (`MediaType.get(CONTENT_TYPE)`), and then included on the request.

**Comment by richardmcclellan** [comment]
Closing in favor of https://github.com/aws-amplify/amplify-android/pull/488

**Pull request closed without merging**